qiskit_machine_learning/algorithms/effective_dimension.py

```diff
@@ -0,0 +1,355 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""An implementation of the effective dimension algorithm."""
+
+import time
+from typing import Optional, Union, List, Callable, Tuple
+
+import numpy as np
+from scipy.special import logsumexp
+
+from qiskit.utils import algorithm_globals
+
+from ..neural_networks import OpflowQNN, NeuralNetwork
+
+
+class EffectiveDimension:
+    """
+    This class computes the global effective dimension for Qiskit
+    :class:`~qiskit_machine_learning.neural_networks.NeuralNetwork`s
+    following the definition used in [1].
+
+    **References**
+        [1]: Abbas et al., The power of quantum neural networks. `The power of QNNs
+            <https://arxiv.org/pdf/2011.00027.pdf>`_
+    """
+
+    def __init__(
+        self,
+        qnn: NeuralNetwork,
+        params: Optional[Union[List[float], np.ndarray, float]] = None,
+        inputs: Optional[Union[List[float], np.ndarray, float]] = None,
+        num_params: int = 1,
+        num_inputs: int = 1,
+        callback: Optional[Callable[[str], None]] = None,
+    ) -> None:
+
+        """
+        Args:
+            qnn: A Qiskit :class:`~qiskit_machine_learning.neural_networks.NeuralNetwork`,
+                with a specific dimension ``(num_weights)`` that will determine the shape of the
+                Fisher Information Matrix ``(num_inputs * num_params, num_weights, num_weights)``
+                used to compute the global effective dimension for a set of ``inputs``,
+                of shape ``(num_inputs, qnn_input_size)``,
+                and ``params``, of shape ``(num_params, num_weights)``.
+            params: An array of neural network parameters (weights), of shape
+                ``(num_params, num_weights)``.
+            inputs: An array of inputs to the neural network, of shape
+                ``(num_inputs, qnn_input_size)``.
+            num_params: If ``params`` is not provided, the algorithm will
+                randomly sample ``num_params`` parameter sets from a
+                uniform distribution. By default, ``num_params = 1``.
+            num_inputs:  If ``inputs`` is not provided, the algorithm will
+                randomly sample ``num_inputs`` input sets
+                from a normal distribution. By default, ``num_inputs = 1``.
```

Comment:
I must admit I find this a bit confusing to sort out. The qnn has `num_inputs` and `num_weights` properties so at times text here is referring to properties of the qnn, other times its a property here say `num_params` and `num_inputs` is ambigious and I think has different meanings. I think here num_params and num_inputs refer to a length/size of arrays that will be randomly created for params and inputs if not supplied.

I wonder whether params could be say `params: l[Union[List[float], np.ndarray, int]] = 1`, where by default, or with an int supplied, will build out a random set. Otherwise via List or array you supply your own. This would remove the need for `num_params` and `num_inputs` here and any confusion that may arise particularly over the latter. Also for params and inputs these are two dimensional array right as in params it an array of different weight values, num_params of them. Is List[float] correct or should it be a List[List[float]] as in I do not see any reshaping if it were a single list. The inputs one says `(num_inputs, qnn_input_size)` which I think is more correctly `(num_inputs, qnn.num_inputs)`

Reply:
@ElePT please, be careful with variable names. I suggest to use `weights`, `num_weights` for weights, `num_inputs` for the number of features, same as in QNNs, `num_samples` for the number of samples in input data. Thus, `(num_inputs, qnn_input_size)` should become `(num_samples, num_inputs)`. I hope, I did not bring more confusion.

Reply:
Although, I'd prefer to rename `num_inputs` to `num_features` in QNNs. Maybe, next time.

Reply:
I personally find it confusing with the qnn haviing num_weights and num_inputs and names here being the same/similar but from what I can see different meanings. From what I see there an array of different values for weights and one for inputs. These have number of different value/samples in them. And are of dimension M x qnn.num_weights and N x qnn.num_inputs. As far as I can see M and N do not need to be the same. Then how to do that, Could it be as I was alluding to just a single field where you either supply your own array or you tell us the size M or N and we build a random array. Whatever we do, as you say we need to be careful with names - if I look at the end of the tutorial (I did not mention it there yet pending any out come from here - it plots 'trained_params` but the text says `We can see from the graph above that the trained weights lead to a higher local effective dimension,` where it talks about weights. The params field here is a set of values for weights right.